existential-vs-universal-quantification/article.md

```diff
@@ -0,0 +1,309 @@
+veryLongFunction :: a -> b -> c -> d -> e -> f -> g -> h -> i -> j
+veryLongFunction = ...
+
+func = veryLongFunction @_ @_ @_ @_ @_ @_ @_ @_ @_ @Integer ...
+```
+Quite long, right?
+However, this can be simplified with explicit `forall`:
+
+```Haskell
+veryLongFunction :: forall j a b c d e f g h i. a -> b -> c -> d -> e -> f -> g -> h -> i -> j
+veryLongFunction = ...
+
+func = veryLongFunction @Integer ...
+```
+
+Since `j` is the first type variable in the declaration of `veryLongFunction`, you can specify only it.
+This is useful not only in such big examples but also when your function requires any of its arguments to be specified on usage.
+
+### Support of other extensions
+
+`ExplicitForAll` shines most when you need to enable other extensions. Some of them just won't work without the `forall` quantifier. :)
+
+The most common one is `ScopedTypeVariables`, which allows you to pass a type variable from an outer function to one defined in a where clause.
+
+```haskell
+example :: a -> [a] -> [a]
+example x rest = pair ++ rest
+  where
+    pair :: [a]
+    pair = [x, x]
+```
+
+This will cause an error without `ScopedTypeVariables` because compiler thinks that the `a` in `example`'s signature and the `a` in `pair`'s are two different types.
+To fix it, you also need to use extension and add a `forall` quantifier to `example` function, as extension won't work without it.
+
+```haskell
+example :: forall a. a -> [a] -> [a]
+```
+You might also notice how it was useful with the `TypeApplication` extension.
+Other extensions that somehow benefit from the usage of `ExplicitForAll` are `LiberalTypeSynonyms`, `RankNTypes`, and many more.
+
+## Existential quantification
+
+As I said earlier, besides universally quantified variables, Haskell also supports existential types.
+But they also use the `forall` keyword.
+
+It can be implemented this way because these two constructions – `(exists x. p x) -> q` and `forall x. (p x -> q)` – are equivalent in terms of first-order predicate logic.
+For a theoretical proof of this statement, you can check this [thread](https://stackoverflow.com/questions/10753073/whats-the-theoretical-basis-for-existential-types?rq=1).
+
+### Usage
+
+In order to proceed further, we'll have to enable the `RankNTypes` extension. This lets us write higher-rank types.
+
+<hr>
+
+**Higher-rank types**
+
+By default, all types are rank-1. In a rank-1 type, a `forall` cannot appear on the left of an arrow (keep in mind that the `->` operator has precedence over `.`)
+
+For example, this function has a rank-1 type:
+
+```haskell
+f :: forall a b. a -> b -> a
+
+-- Or, to make the precedence explicit:
+f :: forall a b. (a -> b -> a)
+```
+
+With the `RankNTypes` extension, we can write rank-2 types – types where `forall` appears to the left of one arrow:
+
+```haskell
+f :: (forall a. Show a => a -> IO ()) -> IO ()
+```
+
+</hr>
+
+Because `forall a` appears to the left of one arrow, `a` is said to be existentially quantified.
```

Comment:
Which is that arrow? 

This description confuses me, it is left of an arrow in rank1types as well.